entrypoint/ocp/ocp.go

```diff
@@ -55,6 +50,11 @@ func ocpBuildClient() error {
 	}
 	log.Info("Host is running as an OpenShift custom strategy builder.")
 
+	// BuildConfigs do not use overlays. And when running in Buildconfig
+	// mode, force running as unprivileged.
+	os.Setenv("COSA_SKIP_OVERLAY", "skip")
+	os.Setenv("FORCE_UNPRIVILEGED", "1")
```

Comment:
I think we should should probably set these on the pod spec.  `Setenv` isn't threadsafe, Java actually did the right thing in disallowing it from the start.  See:

- https://sourceware.org/bugzilla/show_bug.cgi?id=15607
- https://github.com/rust-lang/rust/pull/24741

Reply:
(I still have the scar from my days on the GNOME team at RH where several of us spent hours and hours at one point chasing an intermittent memory corruption bug that was caused by gnome-settings-daemon calling `setenv()` after a thread had been started, the code changed in an innocent-looking refactoring to start a thread before that)

Reply:
Yes, but....this runs before the COSA commands (i.e. we're not spawning threads yet). I went back and forth on. My thinking was to try and do the "better thing." But to concede a point, there is a pod-template for 3.x and 4.x already.

Reply:
Right, the gnome-settings-daemon code was also fine for years until an apparently innocent refactoring later...

Reply:
(Also this is golang, libraries will absoutely spawn background goroutines that can be threads etc.)

Reply:
Look, I understand your old wounds are informing your opinion here.

I have my scars that that are pushing me the other way. I think that putting this in the pod spec is a bad idea. I'm aiming to remove what's in the pod spec because of scalability. Pushing into the pod Spec works if you only care about one or two pipelines. The reason I am arguing for this world is because of the cost of having to change and update the templates is expensive -- I am trying to reduce the times that we have to go to the ART and Multiarch teams and tell them that a change made in COSA requires action on their part when we can address the problem in a cheaper way. 

I even did an audit of the envVar's being set AND read by the vendored code as a due-dilligence. There is no collision or potential collision in the code. 

I'm holding my ground -- and I reject your debug trama as a reason not to set envars. 

As a middle ground compromise, which I think is unneeded at this point, I am now explicitly setting the envVars in []string which is then passed to command.

Reply:
> I even did an audit of the envVar's being set AND read by the vendored code as a due-diligence. There is no collision or potential collision in the existing code.

Right, I should have been more clear - the real problem is when you link C (or C++/Rust) code into the same process, which we aren't doing right now (AFAIK).  But https://github.com/ostreedev/ostree-go exists and it could make sense to use it - and the second we do so, every `setenv()` call introduces the risk of all the `getenv()` calls on the C side returning arbitrary data and hence in the limit arbitrary behavior.  (Think of things like the `http_proxy` variable, plus things like `PATH` telling us where to find binaries etc)

>  As a middle ground compromise, which I think is unneeded at this point, I am now explicitly setting thecmd.Env for the command.

Oh yep that's totally correct as well and a good solution.

>  The reason I am arguing for this world is because of the cost of having to change and update the templates is expensive -- I am trying to reduce the times that we have to go to the ART and Multiarch teams and tell them that a change made in COSA requires action on their part when we can address the problem in a cheaper way.

Don't we control the pod spec generated from the buildconfig?  I wouldn't expect we fork the copy of that per pipeline.

Reply:
Thank you @cgwalters for the clarification on the concerns; context matters, and the added context is really helpful. 

> Don't we control the pod spec generated from the buildconfig?

The first spec, yes. One of the un-obvious aspects of this PR is the setup of secrets based on Kubernetes Labels and well-formed secrets for automated secret discoveries. Using AWS as an example a secret thusly formed:
```
apiVersion: v1
data:
  config: <REDACTED>
  default_region: <BASED64 encoded>
kind: Secret
metadata:
  labels:
       coreos-assembler.coreos.com/secret: aws
  name: mysecret-thats-not-obviously-an-aws-one
type: Opaque
```
could be automatically discovered (so long as the service account had permissions to _read_ the secret) and would then write `/srv/secret/mysecret-thats-not-obviously-an-aws-one/config` and set `AWS_CONFIG` to that file and `AWS_DEFAULT_REGION` to the value of the unencrypted secret. 

For the pod spec, the question is how the pod is created. In the buildConfig situation, `entry` is already pre-created as the pod. 

BUT, your comment makes me realize that I haven't been imaginative _enough_ in my thinking. And I honestly, I can't believe that I didn't think of this before: if the Service Account has edit permissions, then we create a pod with some of this all set up for us and we can work around the restrictions on `/srv` needing to be an overlay and not needing `FORCE_UNPRIVILEGED.`  Okay, Mr @cgwalters, well done.

Reply:
Right, I'd be pretty surprised if the custom build support didn't allow control over the pod as far as volume mounts, environment variables etc.

But another option is to have a wrapper shell script that sets up environment variables before calling the Go entrypoint.